phosphor-auditlog/auditLogMgr.cpp

```diff
@@ -0,0 +1,123 @@
+
+#include "auditLogParser.hpp"
+#include "auditLogUtils.hpp"
+
+#include <fcntl.h>
+#include <libaudit.h>
+
+#include <phosphor-logging/lg2.hpp>
+#include <sdeventplus/event.hpp>
+#include <sdeventplus/source/event.hpp>
+#include <xyz/openbmc_project/Common/File/error.hpp>
+
+#include <cstring>
+#include <string>
+
+namespace phosphor
+{
+namespace auditlog
+{
+
+sdbusplus::message::unix_fd ALManager::parseAuditLog()
+{
+#ifdef AUDITLOG_KEEP_JSONFILE
+    ALParseFile parsedFile("/tmp/auditLog.json");
+#else
+    ALParseFile parsedFile;
+#endif
+
+    // Initialize to parse full audit log
+    ALParseAll auditParser(parsedFile);
+
+    lg2::debug("Method ParseAuditLog filePath={FILEPATH}", "FILEPATH",
+               parsedFile.getPath());
+
+    // Parse all the events
+    auditParser.doParse();
+
+    // Get file descriptor to return
+    auto fd = openParseFD(parsedFile);
+
+    if (fd >= 0)
+    {
+        /* Schedule the fd to be closed by sdbusplus when it sends it back over
+         * D-Bus.
+         */
+        sdeventplus::Event event = sdeventplus::Event::get_default();
+        fdCloseParseSource = std::make_unique<sdeventplus::source::Defer>(
+            event, std::bind(std::mem_fn(&ALManager::closeParseFD), this, fd,
+                             std::placeholders::_1));
+    }
+
+    return fd;
+}
+
+sdbusplus::message::unix_fd ALManager::getAuditLog(uint32_t maxEvents)
+{
+#ifdef AUDITLOG_KEEP_JSONFILE
+    ALParseFile parsedFile("/tmp/auditEntries.json");
+#else
+    ALParseFile parsedFile;
+#endif
+
+    lg2::debug("Method GetAuditLog: {FILE} maxEvents: {MAXCOUNT}", "FILE",
+               parsedFile.getPath(), "MAXCOUNT", maxEvents);
+
+    ALParseLatest auditParser(maxEvents, parsedFile);
+
+    // Parse events up to maxEvents specified
+    auditParser.doParse();
+
+    // Get file descriptor to return
+    auto fd = openParseFD(parsedFile);
+
+    if (fd >= 0)
+    {
+        /* Schedule the fd to be closed by sdbusplus when it sends it back over
+         * D-Bus.
+         */
+        sdeventplus::Event event = sdeventplus::Event::get_default();
+        fdCloseGetSource = std::make_unique<sdeventplus::source::Defer>(
+            event, std::bind(std::mem_fn(&ALManager::closeGetFD), this, fd,
+                             std::placeholders::_1));
+    }
+
+    return fd;
+}
+
+int ALManager::openParseFD(ALParseFile& parsedFile)
+{
+    // Confirm parsed file exists
+    int fd = -1;
+    std::error_code ec;
+
+    fd = open(parsedFile.getPath().c_str(), O_RDONLY | O_NONBLOCK);
+    if (fd == -1)
+    {
+        auto e = errno;
+        lg2::error("Failed to open parsedFile ERRNO={ERRNO} PATH={PATH}",
+                   "ERRNO", e, "PATH", parsedFile.getPath());
+        throw sdbusplus::xyz::openbmc_project::Common::File::Error::Open();
+    }
+
+    lg2::debug("openParseFD(): fd={PARSEFD}", "PARSEFD", fd);
+
+    return fd;
+}
+
+void ALManager::closeParseFD(int fd, sdeventplus::source::EventBase& /*source*/)
```

Comment:
I think you could probably just use a single event source and so a single function across the D-Bus methods?

Reply:
I wondered about this. I was worried about the interleaving of two different method calls and one of them wiping out the memory of the other's fd to close by doing the "reset()" of the pointer. ([std::unique_ptr::reset](https://cplusplus.com/reference/memory/unique_ptr/reset/))

I see in phosphor-debug-collector it checking and just using the existing fd on a [second call to getFileHandle()](https://github.com/openbmc/phosphor-debug-collector/blob/d8291e5e6dd3efd53500bde85ed6aa1d1f974904/dump_entry.cpp#L40).

That doesn't seem quite correct to me either as the original event closing the fd will result in the fd closing for this one also. Seems like it could result in the second one not being able to access the file.

Am I misunderstanding this? Will the event.loop() not process another method call before it processes the closing of the first call?

Reply:
The documentation for it is:

sd_event_add_defer() adds a new event source that will be dispatched instantly, before the event loop goes to sleep again and waits for new events. 

which implies it will get called as soon as the code gets back to the event loop.

That being said, maybe just leave it how it is so nobody has to worry about it.

Reply:
I'm reading that man page the same way. I do think it looks pretty stupid to have two versions of this here. I'll combine them into one and add a comment about the understanding of the defer. Hopefully that will keep anyone from wondering about it.